File: frontend/src/lib/features/concepts/types/concept.ts
import type { LibraryItem } from '$lib/features/library/types/library';

export type Concept = {
  name: string;
  slug: string;
  category: string;
  tags: string[];
  markdown_path: string;
  created: string | null;
  updated: string | null;
  linked_papers_count: number;
  linked_concepts_count: number;
};

export type ConceptCreate = {
  name: string;
  description: string;
  category: string;
  tags: string[];
};

export type ConceptListResponse = {
  concepts: Concept[];
};

export type ConceptDetail = {
  concept: Concept;
  content: string;
  linked_papers: LibraryItem[];
};

export type PaperConceptLinks = {
  concept_names: string[];
};

File: backend/schemas/concept.py
from datetime import datetime

from pydantic import BaseModel

from schemas.library import LibraryItemResponse


class ConceptCreate(BaseModel):
    name: str
    description: str = ""
    category: str = ""
    tags: list[str] = []


class ConceptResponse(BaseModel):
    name: str
    slug: str
    category: str
    tags: list[str]
    markdown_path: str
    created: datetime | None = None
    updated: datetime | None = None
    linked_papers_count: int = 0
    linked_concepts_count: int = 0


class ConceptListResponse(BaseModel):
    concepts: list[ConceptResponse]


class ConceptDetailResponse(BaseModel):
    concept: ConceptResponse
    content: str
    linked_papers: list[LibraryItemResponse]


class PaperConceptLinksUpdate(BaseModel):
    concept_names: list[str]


class PaperConceptLinksResponse(BaseModel):
    concept_names: list[str]
